File: Chronicler/Chronicler/AuthStore.swift
import Foundation
import Combine

@MainActor
final class AuthStore: ObservableObject {
    @Published private(set) var isAuthenticated = false
    @Published private(set) var email: String?

    // Persisted: auto-play the next chapter when one finishes (default off).
    @Published private(set) var autoplayNext: Bool
    // Persisted: boost playback volume beyond 100% (default off).
    @Published private(set) var volumeBoosted: Bool

    let api = APIClient()

    private let tokenKey = "chronicler.token"
    private let emailKey = "chronicler.email"
    private let autoplayKey = "chronicler.autoplay"
    private let volumeBoostKey = "chronicler.volumeBoost"

    init() {
        autoplayNext = UserDefaults.standard.bool(forKey: autoplayKey)
        volumeBoosted = UserDefaults.standard.bool(forKey: volumeBoostKey)
        if let token = UserDefaults.standard.string(forKey: tokenKey) {
            api.token = token
            email = UserDefaults.standard.string(forKey: emailKey)
            isAuthenticated = true
        }
    }

    func setAutoplay(_ v: Bool) {
        autoplayNext = v
        UserDefaults.standard.set(v, forKey: autoplayKey)
    }

    func setVolumeBoost(_ v: Bool) {
        volumeBoosted = v
        UserDefaults.standard.set(v, forKey: volumeBoostKey)
    }

    func setToken(_ token: String, email: String) {
        api.token = token
        self.email = email
        UserDefaults.standard.set(token, forKey: tokenKey)
        UserDefaults.standard.set(email, forKey: emailKey)
        isAuthenticated = true
    }

    func clear() {
        api.token = nil
        email = nil
        UserDefaults.standard.removeObject(forKey: tokenKey)
        UserDefaults.standard.removeObject(forKey: emailKey)
        isAuthenticated = false
    }
}

File: android/app/src/main/kotlin/app/chronicler/AuthStore.kt
package app.chronicler

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

class AuthStore(context: Context) {
    private val prefs = context.getSharedPreferences("chronicler", Context.MODE_PRIVATE)
    val api = ApiClient()

    var isAuthenticated by mutableStateOf(false)
        private set
    var email by mutableStateOf<String?>(null)
        private set

    // Persisted: auto-play the next chapter when one finishes.
    var autoplayNext by mutableStateOf(prefs.getBoolean("autoplay", false))
        private set

    fun setAutoplay(v: Boolean) {
        autoplayNext = v
        prefs.edit().putBoolean("autoplay", v).apply()
    }

    // Persisted: boost playback volume beyond 100%.
    var volumeBoosted by mutableStateOf(prefs.getBoolean("volumeBoost", false))
        private set

    fun setVolumeBoost(v: Boolean) {
        volumeBoosted = v
        prefs.edit().putBoolean("volumeBoost", v).apply()
    }

    // Persisted: archive grid density (1 / 2 / 3 columns).
    var gridColumns by mutableStateOf(prefs.getInt("gridColumns", 3))
        private set

    fun setGridSize(v: Int) {
        gridColumns = v
        prefs.edit().putInt("gridColumns", v).apply()
    }

    // Persisted: the active visual theme (Tesla vs. Steampunk). Stored as the enum name.
    fun loadThemeMode(): ThemeMode =
        runCatching { ThemeMode.valueOf(prefs.getString("theme", null) ?: "") }
            .getOrDefault(ThemeMode.TESLA)

    fun setThemeMode(mode: ThemeMode) {
        Theme.themeMode = mode
        prefs.edit().putString("theme", mode.name).apply()
    }

    init {
        prefs.getString("token", null)?.let {
            api.token = it
            email = prefs.getString("email", null)
            isAuthenticated = true
        }
    }

    fun setToken(token: String, email: String) {
        api.token = token
        this.email = email
        prefs.edit().putString("token", token).putString("email", email).apply()
        isAuthenticated = true
    }

    fun clear() {
        api.token = null
        email = null
        prefs.edit().remove("token").remove("email").apply()
        isAuthenticated = false
    }
}
